packages/Adapters/POD_C/src/DTK_C_API.cpp

```diff
@@ -0,0 +1,246 @@
+
+  auto domain_map = build_contiguous_map(teuchos_comm, src_num);
+  auto range_map = build_contiguous_map(teuchos_comm, tgt_num);
+
+  // Build the actual DTK map operator
+  // TODO: PointCloudOperatorFactory::create(...) should be a static method!
+  auto factory = DataTransferKit::PointCloudOperatorFactory();
+  auto map_operator = factory.create( domain_map,
+                                      range_map,
+                                      parameters );
+
+  // Set it up
+  auto domain_entities = pull_entities(
+    teuchos_comm, src_coord, src_num, src_layout, space_dim );
+  auto range_entities = pull_entities(
+    teuchos_comm, tgt_coord, tgt_num, tgt_layout, space_dim );
+
+  DataTransferKit::BasicGeometryManager domain_manager(
+    teuchos_comm, space_dim, domain_entities() );
+  DataTransferKit::BasicGeometryManager range_manager(
+    teuchos_comm, space_dim, range_entities() );
+
+  map_operator->setup( domain_manager.functionSpace(),
+                       range_manager.functionSpace() );
+
+  // Return an opaque pointer. User is responsible for calling delete_map(...)
+  return static_cast<DTK_Map*>(map_operator.release().get());
+}
+
+
+//----------------------------------------------------------------------------//
+void DTK_Map_apply( DTK_Map*        dtk_map,
+                    double const*   src_field,
+                    DTK_Data_layout src_layout,
+                    double*         tgt_field,
+                    DTK_Data_layout tgt_layout,
+                    int             field_dim,
+                    char const*     options )
+{
+  // Cast the opaque pointer back to a DTK map operator
+  auto map_operator = static_cast<DataTransferKit::MapOperator*>(dtk_map);
+
+  // Get a communicator
+  auto domain_comm = map_operator->getDomainMap()->getComm();
+  int const comm_rank = domain_comm->getRank();
+
+  // Parse the options
+  if (comm_rank == 0)
+    std::cout << "\nDTK_Map_apply called with options:\n"<< options << "\n";
+  std::stringstream ss;
+  ss.str(options);
+  boost::property_tree::ptree ptree;
+  boost::property_tree::read_info(ss, ptree);
+
+  // Wrap the source and target vectors
+  // TODO:
+  // @Stuart: Can you help with that?
+  auto domain_map = map_operator->getDomainMap();
+  DTK_CHECK( domain_map->isContiguous() );
+  unsigned src_local_num = domain_map->getMaxLocalIndex() - domain_map->getMinLocalIndex() + 1;
+  Teuchos::Array<double> range_field = pull_field(
+    src_field, src_local_num, src_layout, field_dim );
+  DTK_CHECK( range_field.size() == src_local_num*field_dim );
+
+  auto range_map = map_operator->getRangeMap();
+  DTK_CHECK( range_map->isContiguous() );
+  unsigned tgt_local_num = range_map->getMaxLocalIndex() - range_map->getMinLocalIndex() + 1;
+  Teuchos::Array<double> domain_field(tgt_local_num*field_dim);
+
+  // TODO: I suspect I need the domain and range managers to build the
+  // vectors...
```

Comment:
We can get the range and domain maps right from the operator - no need to cache additional information here. The entity centered field is the right choice. I'll help you finish this up.